File: python/julia_client/session.py
import asyncio
import atexit
import os
import re
import shlex
import shutil
import subprocess
import tempfile
import time
import uuid
from io import TextIOWrapper
from pathlib import Path

DEFAULT_TIMEOUT = 60.0
DEFAULT_JULIA_ARGS = ("--threads=auto",)
PKG_PATTERN = re.compile(r"\bPkg\.")
TEMP_SESSION_KEY = "__temp__"


class JuliaSession:
    def __init__(
        self,
        env_dir,
        sentinel,
        *,
        is_temp=False,
        is_test=False,
        julia_args=DEFAULT_JULIA_ARGS,
        julia_cmd=None,
        log_file=None,
    ):
        self.env_dir = env_dir
        self.sentinel = sentinel
        self.is_temp = is_temp
        self.is_test = is_test
        self.julia_args = julia_args
        self.julia_cmd = julia_cmd
        self.process: asyncio.subprocess.Process | None = None
        self.lock = asyncio.Lock()
        self._log_file: TextIOWrapper | None = log_file

    @property
    def project_path(self):
        if self.is_test:
            return str(Path(self.env_dir).parent)
        return self.env_dir

    @property
    def init_code(self):
        if self.is_test:
            return "using TestEnv; TestEnv.activate()"
        return None

    async def start(self):
        parts = shlex.split(self.julia_cmd) if self.julia_cmd else ["julia"]
        executable = parts[0]
        remaining = parts[1:]
        if remaining and remaining[0].startswith("+"):
            channel_args = [remaining[0]]
            extra_flags = remaining[1:]
        else:
            channel_args = []
            extra_flags = remaining

        if not os.path.isabs(executable):
            resolved = shutil.which(executable)
            if resolved is None:
                raise RuntimeError(
                    f"'{executable}' not found in PATH. Install Julia from https://julialang.org/downloads/"
                )
            executable = resolved

        cmd = [
            executable,
            *channel_args,
            "-i",
            *self.julia_args,
            *extra_flags,
            f"--project={self.project_path}",
        ]

        self.process = await asyncio.create_subprocess_exec(
            *cmd,
            cwd=self.env_dir,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            limit=64 * 1024 * 1024,
        )

        await self._execute_raw("", timeout=120.0)
        await self._execute_raw("try; using Revise; catch; end", timeout=120.0)
        if self.init_code:
            await self._execute_raw(self.init_code, timeout=None)

    def is_alive(self):
        return self.process is not None and self.process.returncode is None

    async def execute(self, code, timeout):
        async with self.lock:
            if not self.is_alive():
                raise RuntimeError("Julia session has died unexpectedly")
            hex_encoded = code.encode().hex()
            wrapped = (
                f'try; Revise.revise(); catch; end;'
                f'include_string(Main, String(hex2bytes("{hex_encoded}")));'
                f'nothing'
            )
            if self._log_file:
                ts = time.strftime("%H:%M:%S")
                self._log_file.write(f"[{ts}] julia> {code}\n")
                self._log_file.flush()
            output = await self._execute_raw(wrapped, timeout)
            if self._log_file and output:
                self._log_file.write(f"{output}\n\n")
                self._log_file.flush()
            return output

    async def _execute_raw(self, code, timeout):
        assert self.process is not None
        assert self.process.stdin is not None

        sentinel_cmd = (
            f'flush(stderr); write(stdout, "\\n"); println(stdout, "{self.sentinel}"); flush(stdout)'
        )
        payload = code + "\n" + sentinel_cmd + "\n"
        self.process.stdin.write(payload.encode())
        await self.process.stdin.drain()

        lines: list[str] = []

        async def read_until_sentinel():
            while True:
                raw = await self.process.stdout.readline()
                if not raw:
                    collected = "\n".join(lines)
                    raise RuntimeError(
                        f"Julia process died during execution.\nOutput before death:\n{collected}"
                    )
                line = raw.decode().rstrip("\n").rstrip("\r")
                if line == self.sentinel:
                    break
                lines.append(line)
            if lines and lines[-1] == "":
                lines.pop()
            return "\n".join(lines)

        if timeout is not None:
            try:
                return await asyncio.wait_for(read_until_sentinel(), timeout=timeout)
            except asyncio.TimeoutError:
                self.process.kill()
                await self.process.wait()
                partial = "\n".join(lines)
                msg = f"Execution timed out after {timeout}s. Session killed; it will restart on next call."
                if partial:
                    msg += f"\n\nOutput before timeout:\n{partial}"
                raise RuntimeError(msg)
        else:
            return await read_until_sentinel()

    async def kill(self):
        if self.process is not None and self.process.returncode is None:
            self.process.kill()
            await self.process.wait()
        if self.is_temp and os.path.isdir(self.env_dir):
            shutil.rmtree(self.env_dir, ignore_errors=True)


class SessionManager:
    def __init__(self, julia_args=DEFAULT_JULIA_ARGS):
        self.julia_args = julia_args
        self._sessions: dict[str, JuliaSession] = {}
        self._create_locks: dict[str, asyncio.Lock] = {}
        self._global_lock = asyncio.Lock()
        self._log_dir = tempfile.mkdtemp(prefix="julia-client-logs-")
        self._log_files: dict[str, TextIOWrapper] = {}
        atexit.register(self._cleanup_logs)

    def _get_log_file(self, key):
        if key not in self._log_files:
            safe_name = key.replace("/", "_").replace("\\", "_").strip("_") or "temp"
            path = os.path.join(self._log_dir, f"{safe_name}.log")
            self._log_files[key] = open(path, "a")
        return self._log_files[key]

    def _cleanup_logs(self):
        for f in self._log_files.values():
            try:
                f.close()
            except Exception:
                pass
        shutil.rmtree(self._log_dir, ignore_errors=True)

    def _key(self, env_path):
        if env_path is None:
            return TEMP_SESSION_KEY
        return str(Path(env_path).resolve())

    async def get_or_create(self, env_path, julia_cmd=None):
        key = self._key(env_path)

        if key in self._sessions and self._sessions[key].is_alive():
            if self._sessions[key].julia_cmd == julia_cmd:
                return self._sessions[key]
            await self._sessions[key].kill()
            del self._sessions[key]

        async with self._global_lock:
            if key not in self._create_locks:
                self._create_locks[key] = asyncio.Lock()
            create_lock = self._create_locks[key]

        async with create_lock:
            if key in self._sessions and self._sessions[key].is_alive():
                if self._sessions[key].julia_cmd == julia_cmd:
                    return self._sessions[key]
                await self._sessions[key].kill()
                del self._sessions[key]

            if key in self._sessions:
                await self._sessions[key].kill()
                del self._sessions[key]

            sentinel = f"__JULIA_CLIENT_{uuid.uuid4().hex}__"
            is_temp = env_path is None
            if is_temp:
                env_dir = tempfile.mkdtemp(prefix="julia-client-")
                is_test = False
            else:
                resolved = Path(env_path).resolve()
                env_dir = str(resolved)
                is_test = resolved.name == "test"

            session = JuliaSession(
                env_dir, sentinel,
                is_temp=is_temp, is_test=is_test,
                julia_args=self.julia_args,
                julia_cmd=julia_cmd,
                log_file=self._get_log_file(key),
            )
            await session.start()
            self._sessions[key] = session
            return session

    async def restart(self, env_path):
        key = self._key(env_path)
        if key in self._sessions:
            await self._sessions[key].kill()
            del self._sessions[key]

    def list_sessions(self):
        result = []
        for key, session in self._sessions.items():
            info = {
                "env_path": session.env_dir,
                "alive": session.is_alive(),
                "temp": session.is_temp,
            }
            if session.julia_cmd is not None:
                info["julia_cmd"] = session.julia_cmd
            if key in self._log_files:
                info["log_file"] = self._log_files[key].name
            result.append(info)
        return result

    async def shutdown(self):
        for session in self._sessions.values():
            await session.kill()
        self._sessions.clear()
        self._cleanup_logs()

File: go/session.go
package main

import (
	"bufio"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"golang.org/x/sync/singleflight"
)

const (
	defaultEvalTimeout = 60.0
	startupTimeout     = 120.0
)

// JuliaSession manages a single persistent Julia subprocess.
type JuliaSession struct {
	projectVal string // pre-computed --project= arg (also used for display)
	sentinel   string
	juliaCmd   string

	proc   *exec.Cmd
	stdin  io.WriteCloser
	stdout *bufio.Reader
	mu     sync.Mutex

	dead    atomic.Bool
	logFile *os.File
}

func newSentinel() string {
	b := make([]byte, 16)
	rand.Read(b)
	return fmt.Sprintf("__JULIA_CLIENT_%s__", hex.EncodeToString(b))
}

func newJuliaSession(projectVal, sentinel, juliaCmd string, logFile *os.File) *JuliaSession {
	return &JuliaSession{
		projectVal: projectVal,
		sentinel:   sentinel,
		juliaCmd:   juliaCmd,
		logFile:    logFile,
	}
}


func (s *JuliaSession) start(workDir string) error {
	exe := "julia"
	var channelArgs, extraFlags []string

	if s.juliaCmd != "" {
		parts := strings.Fields(s.juliaCmd)
		exe = parts[0]
		rest := parts[1:]
		if len(rest) > 0 && strings.HasPrefix(rest[0], "+") {
			channelArgs = rest[:1]
			extraFlags = rest[1:]
		} else {
			extraFlags = rest
		}
	}

	if !filepath.IsAbs(exe) {
		resolved, err := exec.LookPath(exe)
		if err != nil {
			return fmt.Errorf("'%s' not found in PATH. Install Julia from https://julialang.org/downloads/", exe)
		}
		exe = resolved
	}

	args := append(channelArgs, "-i", "--threads=auto")
	args = append(args, extraFlags...)
	args = append(args, fmt.Sprintf("--project=%s", s.projectVal))

	cmd := exec.Command(exe, args...)
	cmd.Dir = workDir

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}

	// Merge stdout+stderr into a single pipe (mirrors Python's stderr=subprocess.STDOUT)
	pr, pw, err := os.Pipe()
	if err != nil {
		return err
	}
	cmd.Stdout = pw
	cmd.Stderr = pw

	if err := cmd.Start(); err != nil {
		pw.Close()
		pr.Close()
		return err
	}
	pw.Close() // parent only reads; close the write end so EOF propagates when process exits

	s.proc = cmd
	s.stdin = stdin
	s.stdout = bufio.NewReaderSize(pr, 64*1024*1024)

	// Wait for Julia's interactive prompt to appear
	if _, err := s.executeRaw("", startupTimeout); err != nil {
		return fmt.Errorf("Julia startup failed: %w", err)
	}
	// Mirror the Julia REPL: load InteractiveUtils so subtypes, @which, etc. work
	if _, err := s.executeRaw("using InteractiveUtils", startupTimeout); err != nil {
		return fmt.Errorf("failed to load InteractiveUtils: %w", err)
	}
	return nil
}

func (s *JuliaSession) isAlive() bool {
	return !s.dead.Load()
}

type readResult struct {
	output string
	err    error
}

func (s *JuliaSession) executeRaw(code string, timeoutSecs float64) (string, error) {
	// The sentinel command writes an extra "\n" before the marker so it always
	// starts on its own line even when the user code didn't end with a newline.
	// We strip exactly that one "\n" when assembling the result.
	sentinelCmd := fmt.Sprintf(
		"flush(stderr); write(stdout, \"\\n\"); println(stdout, \"%s\"); flush(stdout)\n",
		s.sentinel,
	)
	if _, err := io.WriteString(s.stdin, code+"\n"+sentinelCmd); err != nil {
		return "", err
	}

	ch := make(chan readResult, 1)
	go func() {
		var buf strings.Builder
		for {
			line, err := s.stdout.ReadString('\n')
			if strings.TrimRight(line, "\r\n") == s.sentinel {
				// Strip the one "\n" we injected before the sentinel.
				out := buf.String()
				if len(out) > 0 && out[len(out)-1] == '\n' {
					out = out[:len(out)-1]
				}
				ch <- readResult{out, nil}
				return
			}
			if err != nil {
				s.dead.Store(true)
				ch <- readResult{buf.String(), fmt.Errorf("Julia process died during execution.\nOutput before death:\n%s", buf.String())}
				return
			}
			buf.WriteString(line)
		}
	}()

	if timeoutSecs <= 0 {
		r := <-ch
		return r.output, r.err
	}

	timer := time.NewTimer(time.Duration(float64(time.Second) * timeoutSecs))
	defer timer.Stop()

	select {
	case r := <-ch:
		return r.output, r.err
	case <-timer.C:
		s.proc.Process.Kill()
		s.proc.Wait()
		s.dead.Store(true)
		r := <-ch // goroutine unblocks on EOF after kill
		msg := fmt.Sprintf("Execution timed out after %vs. Session killed; it will restart on next call.", timeoutSecs)
		if r.output != "" {
			msg += "\n\nOutput before timeout:\n" + r.output
		}
		return "", fmt.Errorf("%s", msg)
	}
}

func (s *JuliaSession) execute(code string, timeoutSecs float64, printResult bool) (string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.dead.Load() {
		return "", fmt.Errorf("Julia session has died unexpectedly")
	}

	hexCode := hex.EncodeToString([]byte(code))
	var wrapped string
	if printResult {
		wrapped = fmt.Sprintf(
			`show(IOContext(stdout, :limit => true), MIME("text/plain"), include_string(Main, String(hex2bytes("%s"))));println(stdout)`,
			hexCode,
		)
	} else {
		wrapped = fmt.Sprintf(
			`include_string(Main, String(hex2bytes("%s")));nothing`,
			hexCode,
		)
	}

	if s.logFile != nil {
		fmt.Fprintf(s.logFile, "[%s] julia> %s\n", time.Now().Format("15:04:05"), code)
	}

	output, err := s.executeRaw(wrapped, timeoutSecs)
	if err != nil {
		return "", err
	}
	if s.logFile != nil && output != "" {
		fmt.Fprintf(s.logFile, "%s\n\n", output)
	}
	return output, nil
}

func (s *JuliaSession) kill() {
	s.dead.Store(true)
	if s.proc != nil && s.proc.Process != nil {
		s.proc.Process.Kill()
		s.proc.Wait()
	}
	if s.logFile != nil {
		s.logFile.Close()
	}
}

// SessionManager tracks multiple named Julia sessions.
type SessionManager struct {
	mu       sync.Mutex
	sessions map[string]*JuliaSession
	sf       singleflight.Group
	logDir   string
}

func newSessionManager() *SessionManager {
	logDir, _ := os.MkdirTemp("", "julia-client-logs-")
	return &SessionManager{
		sessions: make(map[string]*JuliaSession),
		logDir:   logDir,
	}
}

// key returns the session map key.
// Priority: explicit session label > explicit project path > cwd.
func (m *SessionManager) key(session, project, cwd string) string {
	if session != "" {
		return "~" + session
	}
	if project != "" && project != "@." {
		abs, _ := filepath.Abs(project)
		return abs
	}
	return cwd
}

func (m *SessionManager) openLogFile(key string) *os.File {
	safe := strings.NewReplacer("/", "_", "\\", "_").Replace(strings.Trim(key, "/~"))
	if safe == "" {
		safe = "default"
	}
	f, _ := os.OpenFile(filepath.Join(m.logDir, safe+".log"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	return f
}

func (m *SessionManager) getOrCreate(cwd, project, session, juliaCmd string) (*JuliaSession, error) {
	key := m.key(session, project, cwd)

	// Fast path: return existing live session without singleflight overhead.
	m.mu.Lock()
	sess := m.sessions[key]
	m.mu.Unlock()
	if sess != nil && sess.isAlive() && sess.juliaCmd == juliaCmd {
		return sess, nil
	}

	// Slow path: deduplicate concurrent creation for the same key.
	v, err, _ := m.sf.Do(key, func() (any, error) {
		m.mu.Lock()
		sess := m.sessions[key]
		m.mu.Unlock()
		if sess != nil && sess.isAlive() && sess.juliaCmd == juliaCmd {
			return sess, nil
		}
		if sess != nil {
			sess.kill()
			m.mu.Lock()
			delete(m.sessions, key)
			m.mu.Unlock()
		}

		projectVal := project
		if projectVal == "" {
			projectVal = "@."
		}
		sess = newJuliaSession(projectVal, newSentinel(), juliaCmd, m.openLogFile(key))
		if err := sess.start(cwd); err != nil {
			return nil, err
		}

		m.mu.Lock()
		m.sessions[key] = sess
		m.mu.Unlock()
		return sess, nil
	})
	if err != nil {
		return nil, err
	}
	return v.(*JuliaSession), nil
}

func (m *SessionManager) remove(session, project, cwd string) {
	key := m.key(session, project, cwd)
	m.mu.Lock()
	delete(m.sessions, key)
	m.mu.Unlock()
}

func (m *SessionManager) restart(session, project, cwd string) {
	key := m.key(session, project, cwd)
	m.mu.Lock()
	sess := m.sessions[key]
	delete(m.sessions, key)
	m.mu.Unlock()
	if sess != nil {
		sess.kill()
	}
}

type sessionInfo struct {
	project  string
	alive    bool
	juliaCmd string
	logFile  string
}

func (m *SessionManager) list() []sessionInfo {
	m.mu.Lock()
	defer m.mu.Unlock()
	result := make([]sessionInfo, 0, len(m.sessions))
	for _, sess := range m.sessions {
		info := sessionInfo{
			project:  sess.projectVal,
			alive:    sess.isAlive(),
			juliaCmd: sess.juliaCmd,
		}
		if sess.logFile != nil {
			info.logFile = sess.logFile.Name()
		}
		result = append(result, info)
	}
	return result
}

func (m *SessionManager) shutdown() {
	m.mu.Lock()
	sessions := make([]*JuliaSession, 0, len(m.sessions))
	for _, s := range m.sessions {
		sessions = append(sessions, s)
	}
	m.sessions = make(map[string]*JuliaSession)
	m.mu.Unlock()

	for _, s := range sessions {
		s.kill()
	}
	os.RemoveAll(m.logDir)
}
